Custom terminal fonts support

ixrock opened the pull request:
New fonts list for terminal with current `fontSize` live-preview: _(fonts downloaded from https://fonts.google.com/)_

<img width="909" alt="Screenshot 2022-05-25 at 16 48 44" src="https://user-images.githubusercontent.com/6377066/170278885-6d58342e-5a53-4962-9d69-ec68d359671c.png">

fix https://github.com/lensapp/lens/issues/5132
fix https://github.com/lensapp/lens/issues/5133

ixrock committed:
custom terminal fonts -- part 1

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
custom terminal fonts -- part 2

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
- added select-list with bundled fonts instead of single text-area (settings -> terminal)
- fix: proper assets exports with webpack (see: fonts.scss)
- fix: lint

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
terminal custom fonts -- fine-tuning

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge branch 'master' into terminal-fonts-customization

ixrock committed:
fix: refresh terminal font after changing in the preferences / keep cluster iframe accessible in DOM while not active/focused

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
update snapshots

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/renderer/components/+preferences/terminal.tsx:
```diff
@@ -88,11 +95,11 @@ const NonInjectedTerminal = observer(({
+            themeName="lens"
             value={userStore.terminalConfig.fontFamily}
-            onChange={(value) => userStore.terminalConfig.fontFamily = value}
+            options={supportedCustomFonts}
```
Minor: It would be cool if the labels for these options were in the font that they represent.

ixrock replied:
Good idea. Gonna take a look soonish. 
_(since all fonts are preloaded and ready to use it's easy-picky addition)_

ixrock replied:
<img width="909" alt="Screenshot 2022-05-25 at 16 48 44" src="https://user-images.githubusercontent.com/6377066/170278648-75ff1e2a-c984-45f5-86a0-aa0957a0c8ff.png">

Nokel81 reviewed (dismissed):
LGTM!

ixrock commented on src/renderer/components/fonts.scss:
```diff
@@ -3,34 +3,66 @@
+// https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/RobotoMono
+@font-face {
+  font-family: "RobotoMono";
+  src: local("RobotoMono"), url("../fonts/roboto-mono-nerd.ttf") format("truetype");
```
Not sure if/when it's needed _(patched font)_ Anybody any clue? 
Could we replace it with normal roboto-mono dowloaded from [google.font](https://fonts.google.com/) ?

Nokel81 replied:
I think that is a @aleksfront question.

aleksfront replied:
This patched font contains some fancy glyphs modern shells can use. So this kind of things are possible with `roboto-mono-nerd`:
<img width="487" alt="shell glyphs" src="https://user-images.githubusercontent.com/9607060/170482771-2b1bff75-6e88-4416-ad03-6fbd60b99240.png">

Other fonts will show blank squares instead of icons/glyphs.

Nokel81 replied:
Could we have those glifs patched in, or maybe as a fallback?

aleksfront replied:
Not sure what you mean, contain glifs separately somehow? Hovewer, as I said before, these fancy shell icons are only possible with patched Roboto.

ixrock commented on src/renderer/components/mixins.scss:
```diff
@@ -59,20 +59,3 @@
-
-// Makes custom @font-family available at earlier stages.
-// Element must exist in DOM as soon as possible to initiate preloading.
-@mixin font-preload($fontFamily) {
```
Seems like overkill since can serve only single element and require duplicating exported font name (declared in `fonts.scss`). More simple solution would be to define font directly via `style="font-family: %ExportedNameInFontsCss%"` _(see: `template.html`)_ and enable early-preloading therefore.

If early-preloading not the point it's just enough to import and bundle custom font in `fonts.scss`

ixrock commented on src/renderer/components/cluster-manager/cluster-frame-handler.ts:
```diff
@@ -113,9 +112,9 @@ export class ClusterFrameHandler {
+        (view: LensView) => {
           logger.info(`[LENS-VIEW]: cluster id=${clusterId} should now be visible`);
-          view.frame.style.display = "flex";
+          view.frame.classList.remove("hidden");
```
_FYI:_ It's always better to move all styles management if possible to CSS files and just toggle classnames in JS.

ixrock commented on src/renderer/components/dock/terminal/terminal.ts:
```diff
@@ -197,12 +204,21 @@ export class Terminal {
+    this.fit();
+
+    // provide css-variable within `:root {}`
+    document.documentElement.style.setProperty("--font-terminal", fontFamily);
```
Exported global "live" css-variable.

ixrock committed:
display terminal's custom font preview with font-name in the select-box + live-preload for current `fontSize`, fix lint

Signed-off-by: Roman <ixrock@gmail.com>

aleksfront reviewed (dismissed):
Need to check iframe styles a bit more:

https://user-images.githubusercontent.com/9607060/170486618-ad3035fd-d658-43f9-833a-ce697a8a6ca2.mov

Also:
1. What if user wants to use font not listed in the dropdown?
2. What was the criterion for choosing these fonts? In short: why no more and no less fonts?
3. Can we somehow list font names from the google fonts and download them on demand? It looks a bit excessively that we contain all mono fonts in the bundle (2.1 MB) but I suggest their usage is very rare.

ixrock committed:
fix: correct positioning with multiple cluster iframes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: bring Roboto as main font back

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge branch 'master' into terminal-fonts-customization

ixrock committed:
fixes for <Input/>:
- remove duplicated error messages for sync validators
- don't propagate invalid values on change (uncontrolled components only)
- more informative error message for numeric input with min/max info

Signed-off-by: Roman <ixrock@gmail.com>

ixrock commented on src/renderer/components/input/input.tsx:
```diff
@@ -173,22 +175,18 @@ export class Input extends React.Component<InputProps, State> {
             error => this.getValidatorError(value, validator) || error,
           ),
         );
-      } else {
-        if (!validator.validate(value, this.props)) {
-          errors.push(this.getValidatorError(value, validator));
-        }
```
This was causing to doubling same error message for sync validators:
![image](https://user-images.githubusercontent.com/6377066/171930477-5ce73335-b448-47a8-a699-ad9c165c1f34.png)

Nokel81 replied:
I thought I also fixed this... Ah it is part of https://github.com/lensapp/lens/pull/5403 which hasn't been merged yet.

jim-docker reviewed (dismissed):
bug I found is gone

ixrock committed:
update snapshots, fix tests

Signed-off-by: Roman <ixrock@gmail.com>

aleksfront reviewed (approved):
LGTM
